source/utilities.cc

```diff
@@ -236,35 +236,80 @@ namespace aspect
       int    wn = 0;    // the  winding number counter
       int   j=pointNo-1;
 
-
       // loop through all edges of the polygon
       for (int i=0; i<pointNo; i++)
         {
           // edge from V[i] to  V[i+1]
           if (point_list[j][1] <= point[1])
             {
               // start y <= P.y
-              if (point_list[i][1]  >= point[1])      // an upward crossing
-                if (( (point_list[i][0] - point_list[j][0]) * (point[1] - point_list[j][1])
-                      - (point[0] -  point_list[j][0]) * (point_list[i][1] - point_list[j][1]) ) >= 0)
-                  {
-                    // P left of  edge
-                    ++wn;            // have  a valid up intersect
-                  }
+              if (point_list[i][1] >= point[1])      // an upward crossing
+                {
+                  const double is_left = (point_list[i][0] - point_list[j][0]) * (point[1] - point_list[j][1])
+                                         - (point[0] -  point_list[j][0]) * (point_list[i][1] - point_list[j][1]);
+
+                  if ( is_left > 0 && point_list[i][1] > point[1])
+                    {
+                      // P left of  edge
+                      ++wn;            // have  a valid up intersect
+                    }
+                  else if ( is_left == 0)
+                    {
+                      // The point is exactly on the infinite line.
+                      // determine if it is on the segment
+                      const double dot_product = (point[0] - point_list[j][0]) * (point_list[i][0] - point_list[j][0])
+                                                 + (point[1] - point_list[j][1]) * (point_list[i][1] - point_list[j][1]);
```

Comment:
Oh, I misread then. But you can write it as `(point - point_list[j])*(point_list[i] - point_list[j])` then.